src/ray/core_worker/transport/direct_task_transport.cc

```diff
@@ -0,0 +1,189 @@
+    auto count = task.ArgIdCount(i);
+    if (count > 0) {
+      const auto &id = task.ArgId(i, 0);
+      if (id == obj_id) {
+        auto *mutable_arg = msg.mutable_args(i);
+        mutable_arg->clear_object_ids();
+        if (value->HasData()) {
+          const auto &data = value->GetData();
+          mutable_arg->set_data(data->Data(), data->Size());
+        }
+        if (value->HasMetadata()) {
+          const auto &metadata = value->GetMetadata();
+          mutable_arg->set_metadata(metadata->Data(), metadata->Size());
+        }
+        found = true;
+      }
+    }
+  }
+  RAY_CHECK(found) << "obj id " << obj_id << " not found";
+}
+
+void LocalDependencyResolver::ResolveDependencies(const TaskSpecification &task,
+                                                  std::function<void()> on_complete) {
+  absl::flat_hash_set<ObjectID> local_dependencies;
+  for (size_t i = 0; i < task.NumArgs(); i++) {
+    auto count = task.ArgIdCount(i);
+    if (count > 0) {
+      RAY_CHECK(count <= 1) << "multi args not implemented";
+      const auto &id = task.ArgId(i, 0);
+      if (id.IsDirectCallType()) {
+        local_dependencies.insert(id);
+      }
+    }
+  }
+  if (local_dependencies.empty()) {
+    on_complete();
+    return;
+  }
+
+  // This is deleted when the last dependency fetch callback finishes.
+  std::shared_ptr<TaskState> state =
+      std::shared_ptr<TaskState>(new TaskState{task, std::move(local_dependencies)});
+  num_pending_ += 1;
+
+  for (const auto &obj_id : state->local_dependencies) {
+    in_memory_store_.GetAsync(
+        obj_id, [this, state, obj_id, on_complete](std::shared_ptr<RayObject> obj) {
+          RAY_CHECK(obj != nullptr);
+          bool complete = false;
+          {
+            absl::MutexLock lock(&mu_);
+            state->local_dependencies.erase(obj_id);
+            DoInlineObjectValue(obj_id, obj, state->task);
+            if (state->local_dependencies.empty()) {
+              complete = true;
+              num_pending_ -= 1;
+            }
+          }
+          if (complete) {
+            on_complete();
+          }
+        });
+  }
+}
+
+Status CoreWorkerDirectTaskSubmitter::SubmitTask(const TaskSpecification &task_spec) {
+  resolver_.ResolveDependencies(task_spec, [this, task_spec]() {
+    // TODO(ekl) should have a queue per distinct resource type required
+    absl::MutexLock lock(&mu_);
+    RequestNewWorkerIfNeeded(task_spec);
+    queued_tasks_.push_back(task_spec);
+    // The task is now queued and will be picked up by the next leased or newly
+    // idle worker. We are guaranteed a worker will show up since we called
+    // RequestNewWorkerIfNeeded() earlier while holding mu_.
+  });
+  return Status::OK();
+}
+
+void CoreWorkerDirectTaskSubmitter::HandleWorkerLeaseGranted(const std::string &address,
+                                                             int port) {
+  WorkerAddress addr = std::make_pair(address, port);
```

Comment:
Could we make signature of this function take in a `WorkerAddress`?

Reply:
Fixing.